cycle detected when optimizing MIR

Only under `--release`, after https://github.com/rust-lang/rust/pull/142625 https://github.com/EFanZh/ast-demangle no longer builds with a cycle error during inline evaluation. Default `--debug` builds do not hit this.
<!--
Thank you for filing a regression report! 🐛 A regression is something that changed between versions of Rust but was not supposed to.

Please provide a short summary of the regression, along with any information you feel is relevant to replicate it.
-->

### Code

https://github.com/EFanZh/ast-demangle (as of commit b4b0d5d, which is current HEAD)

I expected to see this happen: pass

Instead, this happened: cycle error

```
error[E0391]: cycle detected when optimizing MIR for `rust_v0::<impl at src/rust_v0/mod.rs:60:32: 60:41>::eq`
   --> src/rust_v0/mod.rs:60:32
    |
60  | #[derive(Clone, Eq, Hash, Ord, PartialEq, PartialOrd)]
    |                                ^^^^^^^^^
    |
note: ...which requires optimizing MIR for `rust_v0::<impl at src/rust_v0/mod.rs:210:39: 210:48>::eq`...
   --> src/rust_v0/mod.rs:210:39
    |
210 | #[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
    |                                       ^^^^^^^^^
    = note: ...which again requires optimizing MIR for `rust_v0::<impl at src/rust_v0/mod.rs:60:32: 60:41>::eq`, completing the cycle
note: cycle used when optimizing MIR for `rust_v0::<impl at src/rust_v0/mod.rs:15:39: 15:48>::eq`
   --> src/rust_v0/mod.rs:15:39
    |
15  | #[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
    |                                       ^^^^^^^^^
    = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
```

### Version it worked on

Before https://github.com/rust-lang/rust/pull/142625

### Version with regression

https://github.com/rust-lang/rust/pull/142625

### Backtrace
<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

```
thread 'rustc' panicked at compiler/rustc_errors/src/lib.rs:1838:17:
aborting due to `-Z treat-err-as-bug=1`
stack backtrace:
   0:     0x71ada4175aa3 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h4b84b18f52d871f7
   1:     0x71ada4802c7f - core::fmt::write::hddeac91703b6499d
   2:     0x71ada416b5d3 - std::io::Write::write_fmt::h28f31377e98688f6
   3:     0x71ada4175902 - std::sys::backtrace::BacktraceLock::print::h52d0fd2be6990145
   4:     0x71ada417939a - std::panicking::default_hook::{{closure}}::h77b4c5e7cec39435
   5:     0x71ada4178f1f - std::panicking::default_hook::h61df2e12d1f260cf
   6:     0x71ada3223d33 - std[c539802e89e8d81a]::panicking::update_hook::<alloc[247485d20f619618]::boxed::Box<rustc_driver_impl[3a492dd9a06cbf2d]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x71ada4179beb - std::panicking::rust_panic_with_hook::h2deddd251118f4d3
   8:     0x71ada41798b6 - std::panicking::begin_panic_handler::{{closure}}::h5fe965e188abe62f
   9:     0x71ada4175f69 - std::sys::backtrace::__rust_end_short_backtrace::h0f72bacaad7bd6db
  10:     0x71ada41795cd - __rustc[38c938b2c9dd7d5c]::rust_begin_unwind
  11:     0x71ada08ba720 - core::panicking::panic_fmt::h80892ba21db80006
  12:     0x71ada3274037 - <rustc_errors[793135f862eb652b]::DiagCtxtInner>::panic_if_treat_err_as_bug
  13:     0x71ada5d31e88 - <rustc_errors[793135f862eb652b]::DiagCtxtInner>::emit_diagnostic::{closure#3}
  14:     0x71ada5d2f98d - rustc_interface[b327675518afe1b5]::callbacks::track_diagnostic::<core[cdc97c9c0b6fd462]::option::Option<rustc_span[dd9d35b931de6930]::ErrorGuaranteed>>::{closure#0}
  15:     0x71ada5d2f32b - rustc_interface[b327675518afe1b5]::callbacks::track_diagnostic::<core[cdc97c9c0b6fd462]::option::Option<rustc_span[dd9d35b931de6930]::ErrorGuaranteed>>
  16:     0x71ada5d2e93d - <rustc_errors[793135f862eb652b]::DiagCtxtInner>::emit_diagnostic
  17:     0x71ada5d2e7ff - <rustc_errors[793135f862eb652b]::DiagCtxtHandle>::emit_diagnostic
  18:     0x71ada0fdc5d0 - <rustc_span[dd9d35b931de6930]::ErrorGuaranteed as rustc_errors[793135f862eb652b]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  19:     0x71ada3b9bf8c - rustc_query_system[8d89d55943c1ebd2]::query::plumbing::mk_cycle::<rustc_query_impl[3a213cbfc7e81906]::DynamicConfig<rustc_query_system[8d89d55943c1ebd2]::query::caches::DefaultCache<rustc_type_ir[4c29bc3dc42f1408]::predicate::TraitRef<rustc_middle[5796165e46beff11]::ty::context::TyCtxt>, rustc_middle[5796165e46beff11]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3a213cbfc7e81906]::plumbing::QueryCtxt>
  20:     0x71ada3b82859 - rustc_query_system[8d89d55943c1ebd2]::query::plumbing::cycle_error::<rustc_query_impl[3a213cbfc7e81906]::DynamicConfig<rustc_query_system[8d89d55943c1ebd2]::query::caches::DefIdCache<rustc_middle[5796165e46beff11]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3a213cbfc7e81906]::plumbing::QueryCtxt>
  21:     0x71ada483849a - rustc_query_system[8d89d55943c1ebd2]::query::plumbing::try_execute_query::<rustc_query_impl[3a213cbfc7e81906]::DynamicConfig<rustc_query_system[8d89d55943c1ebd2]::query::caches::DefIdCache<rustc_middle[5796165e46beff11]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3a213cbfc7e81906]::plumbing::QueryCtxt, false>
  22:     0x71ada483711b - rustc_query_impl[3a213cbfc7e81906]::query_impl::optimized_mir::get_query_non_incr::__rust_end_short_backtrace
  23:     0x71ada1a5e420 - <rustc_middle[5796165e46beff11]::ty::context::TyCtxt>::instance_mir
  24:     0x71ada542e9f5 - rustc_mir_transform[b3fbc0bf8a9d046a]::inline::process_blocks::<rustc_mir_transform[b3fbc0bf8a9d046a]::inline::NormalInliner>
  25:     0x71ada542d83f - rustc_mir_transform[b3fbc0bf8a9d046a]::inline::process_blocks::<rustc_mir_transform[b3fbc0bf8a9d046a]::inline::NormalInliner>
  26:     0x71ada542d83f - rustc_mir_transform[b3fbc0bf8a9d046a]::inline::process_blocks::<rustc_mir_transform[b3fbc0bf8a9d046a]::inline::NormalInliner>
  27:     0x71ada542d83f - rustc_mir_transform[b3fbc0bf8a9d046a]::inline::process_blocks::<rustc_mir_transform[b3fbc0bf8a9d046a]::inline::NormalInliner>
  28:     0x71ada547e3fd - <rustc_mir_transform[b3fbc0bf8a9d046a]::inline::Inline as rustc_mir_transform[b3fbc0bf8a9d046a]::pass_manager::MirPass>::run_pass
  29:     0x71ada4810158 - rustc_mir_transform[b3fbc0bf8a9d046a]::run_optimization_passes
  30:     0x71ada51f90a0 - rustc_mir_transform[b3fbc0bf8a9d046a]::optimized_mir
  31:     0x71ada51f8875 - rustc_query_impl[3a213cbfc7e81906]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3a213cbfc7e81906]::query_impl::optimized_mir::dynamic_query::{closure#2}::{closure#0}, rustc_middle[5796165e46beff11]::query::erase::Erased<[u8; 8usize]>>
  32:     0x71ada4837c25 - rustc_query_system[8d89d55943c1ebd2]::query::plumbing::try_execute_query::<rustc_query_impl[3a213cbfc7e81906]::DynamicConfig<rustc_query_system[8d89d55943c1ebd2]::query::caches::DefIdCache<rustc_middle[5796165e46beff11]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3a213cbfc7e81906]::plumbing::QueryCtxt, false>
  33:     0x71ada483711b - rustc_query_impl[3a213cbfc7e81906]::query_impl::optimized_mir::get_query_non_incr::__rust_end_short_backtrace
  34:     0x71ada1a5e420 - <rustc_middle[5796165e46beff11]::ty::context::TyCtxt>::instance_mir
  35:     0x71ada542e9f5 - rustc_mir_transform[b3fbc0bf8a9d046a]::inline::process_blocks::<rustc_mir_transform[b3fbc0bf8a9d046a]::inline::NormalInliner>
  36:     0x71ada542d83f - rustc_mir_transform[b3fbc0bf8a9d046a]::inline::process_blocks::<rustc_mir_transform[b3fbc0bf8a9d046a]::inline::NormalInliner>
  37:     0x71ada542d83f - rustc_mir_transform[b3fbc0bf8a9d046a]::inline::process_blocks::<rustc_mir_transform[b3fbc0bf8a9d046a]::inline::NormalInliner>
  38:     0x71ada542d83f - rustc_mir_transform[b3fbc0bf8a9d046a]::inline::process_blocks::<rustc_mir_transform[b3fbc0bf8a9d046a]::inline::NormalInliner>
  39:     0x71ada547e3fd - <rustc_mir_transform[b3fbc0bf8a9d046a]::inline::Inline as rustc_mir_transform[b3fbc0bf8a9d046a]::pass_manager::MirPass>::run_pass
  40:     0x71ada4810158 - rustc_mir_transform[b3fbc0bf8a9d046a]::run_optimization_passes
  41:     0x71ada51f90a0 - rustc_mir_transform[b3fbc0bf8a9d046a]::optimized_mir
  42:     0x71ada51f8875 - rustc_query_impl[3a213cbfc7e81906]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3a213cbfc7e81906]::query_impl::optimized_mir::dynamic_query::{closure#2}::{closure#0}, rustc_middle[5796165e46beff11]::query::erase::Erased<[u8; 8usize]>>
  43:     0x71ada4837c25 - rustc_query_system[8d89d55943c1ebd2]::query::plumbing::try_execute_query::<rustc_query_impl[3a213cbfc7e81906]::DynamicConfig<rustc_query_system[8d89d55943c1ebd2]::query::caches::DefIdCache<rustc_middle[5796165e46beff11]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3a213cbfc7e81906]::plumbing::QueryCtxt, false>
  44:     0x71ada483711b - rustc_query_impl[3a213cbfc7e81906]::query_impl::optimized_mir::get_query_non_incr::__rust_end_short_backtrace
  45:     0x71ada1a5e420 - <rustc_middle[5796165e46beff11]::ty::context::TyCtxt>::instance_mir
  46:     0x71ada542e9f5 - rustc_mir_transform[b3fbc0bf8a9d046a]::inline::process_blocks::<rustc_mir_transform[b3fbc0bf8a9d046a]::inline::NormalInliner>
  47:     0x71ada542d83f - rustc_mir_transform[b3fbc0bf8a9d046a]::inline::process_blocks::<rustc_mir_transform[b3fbc0bf8a9d046a]::inline::NormalInliner>
  48:     0x71ada542d83f - rustc_mir_transform[b3fbc0bf8a9d046a]::inline::process_blocks::<rustc_mir_transform[b3fbc0bf8a9d046a]::inline::NormalInliner>
  49:     0x71ada547e3fd - <rustc_mir_transform[b3fbc0bf8a9d046a]::inline::Inline as rustc_mir_transform[b3fbc0bf8a9d046a]::pass_manager::MirPass>::run_pass
  50:     0x71ada4810158 - rustc_mir_transform[b3fbc0bf8a9d046a]::run_optimization_passes
  51:     0x71ada51f90a0 - rustc_mir_transform[b3fbc0bf8a9d046a]::optimized_mir
  52:     0x71ada51f8875 - rustc_query_impl[3a213cbfc7e81906]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3a213cbfc7e81906]::query_impl::optimized_mir::dynamic_query::{closure#2}::{closure#0}, rustc_middle[5796165e46beff11]::query::erase::Erased<[u8; 8usize]>>
  53:     0x71ada4837c25 - rustc_query_system[8d89d55943c1ebd2]::query::plumbing::try_execute_query::<rustc_query_impl[3a213cbfc7e81906]::DynamicConfig<rustc_query_system[8d89d55943c1ebd2]::query::caches::DefIdCache<rustc_middle[5796165e46beff11]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3a213cbfc7e81906]::plumbing::QueryCtxt, false>
  54:     0x71ada483711b - rustc_query_impl[3a213cbfc7e81906]::query_impl::optimized_mir::get_query_non_incr::__rust_end_short_backtrace
  55:     0x71ada534af1d - <rustc_metadata[e2b15c2c3f6f00d0]::rmeta::encoder::EncodeContext>::encode_crate_root
  56:     0x71ada5c7d5f0 - <rustc_metadata[e2b15c2c3f6f00d0]::rmeta::encoder::encode_metadata::{closure#3} as core[cdc97c9c0b6fd462]::ops::function::FnOnce<(rustc_middle[5796165e46beff11]::ty::context::TyCtxt, &std[c539802e89e8d81a]::path::Path)>>::call_once
  57:     0x71ada5c7a830 - rustc_metadata[e2b15c2c3f6f00d0]::rmeta::encoder::encode_metadata
  58:     0x71ada575ff98 - rustc_metadata[e2b15c2c3f6f00d0]::fs::encode_and_write_metadata
  59:     0x71ada5b0b68a - <rustc_interface[b327675518afe1b5]::queries::Linker>::codegen_and_build_linker
  60:     0x71ada5b091c7 - rustc_interface[b327675518afe1b5]::passes::create_and_enter_global_ctxt::<core[cdc97c9c0b6fd462]::option::Option<rustc_interface[b327675518afe1b5]::queries::Linker>, rustc_driver_impl[3a492dd9a06cbf2d]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  61:     0x71ada5aed6e4 - rustc_interface[b327675518afe1b5]::interface::run_compiler::<(), rustc_driver_impl[3a492dd9a06cbf2d]::run_compiler::{closure#0}>::{closure#1}
  62:     0x71ada59d29f8 - std[c539802e89e8d81a]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[b327675518afe1b5]::util::run_in_thread_with_globals<rustc_interface[b327675518afe1b5]::util::run_in_thread_pool_with_globals<rustc_interface[b327675518afe1b5]::interface::run_compiler<(), rustc_driver_impl[3a492dd9a06cbf2d]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  63:     0x71ada59d26d6 - <<std[c539802e89e8d81a]::thread::Builder>::spawn_unchecked_<rustc_interface[b327675518afe1b5]::util::run_in_thread_with_globals<rustc_interface[b327675518afe1b5]::util::run_in_thread_pool_with_globals<rustc_interface[b327675518afe1b5]::interface::run_compiler<(), rustc_driver_impl[3a492dd9a06cbf2d]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[cdc97c9c0b6fd462]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  64:     0x71ada59d0cfd - std::sys::pal::unix::thread::Thread::new::thread_start::h7f3d8f76868ca167
  65:     0x71ad9f29caa4 - start_thread
                               at ./nptl/pthread_create.c:447:8
  66:     0x71ad9f329c3c - clone3
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:78:0
  67:                0x0 - <unknown>
```

</p>
</details>

<!--
If you know when this regression occurred, please add a line like below, replacing `{channel}` with one of stable, beta, or nightly.

@rustbot modify labels: +regression-from-stable-to-{channel} -regression-untriaged
-->

--- a/compiler/rustc_mir_transform/src/inline/cycle.rs
+++ b/compiler/rustc_mir_transform/src/inline/cycle.rs
@@ -64,15 +64,15 @@ fn process<'tcx>(
     typing_env: ty::TypingEnv<'tcx>,
     caller: ty::Instance<'tcx>,
     target: LocalDefId,
-    seen: &mut FxHashSet<ty::Instance<'tcx>>,
+    seen: &mut FxHashMap<ty::Instance<'tcx>, bool>,
     involved: &mut FxHashSet<LocalDefId>,
     recursion_limiter: &mut FxHashMap<DefId, usize>,
     recursion_limit: Limit,
 ) -> bool {
     trace!(%caller);
-    let mut cycle_found = false;
+    let mut reaches_root = false;
 
-    for &(callee, args) in tcx.mir_inliner_callees(caller.def) {
+    for &(callee_def_id, args) in tcx.mir_inliner_callees(caller.def) {
         let Ok(args) = caller.try_instantiate_mir_and_normalize_erasing_regions(
             tcx,
             typing_env,
@@ -81,14 +81,17 @@ fn process<'tcx>(
             trace!(?caller, ?typing_env, ?args, "cannot normalize, skipping");
             continue;
         };
-        let Ok(Some(callee)) = ty::Instance::try_resolve(tcx, typing_env, callee, args) else {
-            trace!(?callee, "cannot resolve, skipping");
+        let Ok(Some(callee)) = ty::Instance::try_resolve(tcx, typing_env, callee_def_id, args)
+        else {
+            trace!(?callee_def_id, "cannot resolve, skipping");
             continue;
         };
 
         // Found a path.
         if callee.def_id() == target.to_def_id() {
-            cycle_found = true;
+            reaches_root = true;
+            seen.insert(callee, true);
+            continue;
         }
 
         if tcx.is_constructor(callee.def_id()) {
@@ -101,10 +104,17 @@ fn process<'tcx>(
             continue;
         }
 
-        if seen.insert(callee) {
+        let callee_reaches_root = if let Some(&c) = seen.get(&callee) {
+            // Even if we have seen this callee before, and thus don't need
+            // to recurse into it, we still need to propagate whether it reaches
+            // the root so that we can mark all the involved callers, in case we
+            // end up reaching that same recursive callee through some *other* cycle.
+            c
+        } else {
+            seen.insert(callee, false);
             let recursion = recursion_limiter.entry(callee.def_id()).or_default();
             trace!(?callee, recursion = *recursion);
-            let found_recursion = if recursion_limit.value_within_limit(*recursion) {
+            let callee_reaches_root = if recursion_limit.value_within_limit(*recursion) {
                 *recursion += 1;
                 ensure_sufficient_stack(|| {
                     process(
@@ -122,17 +132,19 @@ fn process<'tcx>(
                 // Pessimistically assume that there could be recursion.
                 true
             };
-            if found_recursion {
-                if let Some(callee) = callee.def_id().as_local() {
-                    // Calling `optimized_mir` of a non-local definition cannot cycle.
-                    involved.insert(callee);
-                }
-                cycle_found = true;
+            seen.insert(callee, callee_reaches_root);
+            callee_reaches_root
+        };
+        if callee_reaches_root {
+            if let Some(callee_def_id) = callee.def_id().as_local() {
+                // Calling `optimized_mir` of a non-local definition cannot cycle.
+                involved.insert(callee_def_id);
             }
+            reaches_root = true;
         }
     }
 
-    cycle_found
+    reaches_root
 }
 
 #[instrument(level = "debug", skip(tcx), ret)]
@@ -166,7 +178,7 @@ pub(crate) fn mir_callgraph_cyclic<'tcx>(
         typing_env,
         root_instance,
         root,
-        &mut FxHashSet::default(),
+        &mut FxHashMap::default(),
         &mut involved,
         &mut FxHashMap::default(),
         recursion_limit,

--- a/tests/mir-opt/inline_double_cycle.a.Inline.panic-abort.diff
+++ b/tests/mir-opt/inline_double_cycle.a.Inline.panic-abort.diff
@@ -1,0 +1,48 @@
+- // MIR for `a` before Inline
++ // MIR for `a` after Inline
+  
+  fn a() -> () {
+      let mut _0: ();
+      let _1: ();
+      let mut _2: ();
+      let _3: ();
+      let mut _4: ();
++     let mut _5: fn() {a};
++     let mut _6: fn() {b};
++     scope 1 (inlined <fn() {a} as FnOnce<()>>::call_once - shim(fn() {a})) {
++     }
++     scope 2 (inlined <fn() {b} as FnOnce<()>>::call_once - shim(fn() {b})) {
++     }
+  
+      bb0: {
+          StorageLive(_1);
+          StorageLive(_2);
+          _2 = ();
+-         _1 = <fn() {a} as FnOnce<()>>::call_once(a, move _2) -> [return: bb1, unwind unreachable];
++         StorageLive(_5);
++         _5 = a;
++         _1 = move _5() -> [return: bb1, unwind unreachable];
+      }
+  
+      bb1: {
++         StorageDead(_5);
+          StorageDead(_2);
+          StorageDead(_1);
+          StorageLive(_3);
+          StorageLive(_4);
+          _4 = ();
+-         _3 = <fn() {b} as FnOnce<()>>::call_once(b, move _4) -> [return: bb2, unwind unreachable];
++         StorageLive(_6);
++         _6 = b;
++         _3 = move _6() -> [return: bb2, unwind unreachable];
+      }
+  
+      bb2: {
++         StorageDead(_6);
+          StorageDead(_4);
+          StorageDead(_3);
+          _0 = const ();
+          return;
+      }
+  }
+  

--- a/tests/mir-opt/inline_double_cycle.a.Inline.panic-unwind.diff
+++ b/tests/mir-opt/inline_double_cycle.a.Inline.panic-unwind.diff
@@ -1,0 +1,48 @@
+- // MIR for `a` before Inline
++ // MIR for `a` after Inline
+  
+  fn a() -> () {
+      let mut _0: ();
+      let _1: ();
+      let mut _2: ();
+      let _3: ();
+      let mut _4: ();
++     let mut _5: fn() {a};
++     let mut _6: fn() {b};
++     scope 1 (inlined <fn() {a} as FnOnce<()>>::call_once - shim(fn() {a})) {
++     }
++     scope 2 (inlined <fn() {b} as FnOnce<()>>::call_once - shim(fn() {b})) {
++     }
+  
+      bb0: {
+          StorageLive(_1);
+          StorageLive(_2);
+          _2 = ();
+-         _1 = <fn() {a} as FnOnce<()>>::call_once(a, move _2) -> [return: bb1, unwind continue];
++         StorageLive(_5);
++         _5 = a;
++         _1 = move _5() -> [return: bb1, unwind continue];
+      }
+  
+      bb1: {
++         StorageDead(_5);
+          StorageDead(_2);
+          StorageDead(_1);
+          StorageLive(_3);
+          StorageLive(_4);
+          _4 = ();
+-         _3 = <fn() {b} as FnOnce<()>>::call_once(b, move _4) -> [return: bb2, unwind continue];
++         StorageLive(_6);
++         _6 = b;
++         _3 = move _6() -> [return: bb2, unwind continue];
+      }
+  
+      bb2: {
++         StorageDead(_6);
+          StorageDead(_4);
+          StorageDead(_3);
+          _0 = const ();
+          return;
+      }
+  }
+  

--- a/tests/mir-opt/inline_double_cycle.b.Inline.panic-abort.diff
+++ b/tests/mir-opt/inline_double_cycle.b.Inline.panic-abort.diff
@@ -1,0 +1,48 @@
+- // MIR for `b` before Inline
++ // MIR for `b` after Inline
+  
+  fn b() -> () {
+      let mut _0: ();
+      let _1: ();
+      let mut _2: ();
+      let _3: ();
+      let mut _4: ();
++     let mut _5: fn() {b};
++     let mut _6: fn() {a};
++     scope 1 (inlined <fn() {b} as FnOnce<()>>::call_once - shim(fn() {b})) {
++     }
++     scope 2 (inlined <fn() {a} as FnOnce<()>>::call_once - shim(fn() {a})) {
++     }
+  
+      bb0: {
+          StorageLive(_1);
+          StorageLive(_2);
+          _2 = ();
+-         _1 = <fn() {b} as FnOnce<()>>::call_once(b, move _2) -> [return: bb1, unwind unreachable];
++         StorageLive(_5);
++         _5 = b;
++         _1 = move _5() -> [return: bb1, unwind unreachable];
+      }
+  
+      bb1: {
++         StorageDead(_5);
+          StorageDead(_2);
+          StorageDead(_1);
+          StorageLive(_3);
+          StorageLive(_4);
+          _4 = ();
+-         _3 = <fn() {a} as FnOnce<()>>::call_once(a, move _4) -> [return: bb2, unwind unreachable];
++         StorageLive(_6);
++         _6 = a;
++         _3 = move _6() -> [return: bb2, unwind unreachable];
+      }
+  
+      bb2: {
++         StorageDead(_6);
+          StorageDead(_4);
+          StorageDead(_3);
+          _0 = const ();
+          return;
+      }
+  }
+  

--- a/tests/mir-opt/inline_double_cycle.b.Inline.panic-unwind.diff
+++ b/tests/mir-opt/inline_double_cycle.b.Inline.panic-unwind.diff
@@ -1,0 +1,48 @@
+- // MIR for `b` before Inline
++ // MIR for `b` after Inline
+  
+  fn b() -> () {
+      let mut _0: ();
+      let _1: ();
+      let mut _2: ();
+      let _3: ();
+      let mut _4: ();
++     let mut _5: fn() {b};
++     let mut _6: fn() {a};
++     scope 1 (inlined <fn() {b} as FnOnce<()>>::call_once - shim(fn() {b})) {
++     }
++     scope 2 (inlined <fn() {a} as FnOnce<()>>::call_once - shim(fn() {a})) {
++     }
+  
+      bb0: {
+          StorageLive(_1);
+          StorageLive(_2);
+          _2 = ();
+-         _1 = <fn() {b} as FnOnce<()>>::call_once(b, move _2) -> [return: bb1, unwind continue];
++         StorageLive(_5);
++         _5 = b;
++         _1 = move _5() -> [return: bb1, unwind continue];
+      }
+  
+      bb1: {
++         StorageDead(_5);
+          StorageDead(_2);
+          StorageDead(_1);
+          StorageLive(_3);
+          StorageLive(_4);
+          _4 = ();
+-         _3 = <fn() {a} as FnOnce<()>>::call_once(a, move _4) -> [return: bb2, unwind continue];
++         StorageLive(_6);
++         _6 = a;
++         _3 = move _6() -> [return: bb2, unwind continue];
+      }
+  
+      bb2: {
++         StorageDead(_6);
+          StorageDead(_4);
+          StorageDead(_3);
+          _0 = const ();
+          return;
+      }
+  }
+  

--- a/tests/mir-opt/inline_double_cycle.rs
+++ b/tests/mir-opt/inline_double_cycle.rs
@@ -1,0 +1,22 @@
+// EMIT_MIR_FOR_EACH_PANIC_STRATEGY
+// skip-filecheck
+//@ test-mir-pass: Inline
+//@ edition: 2021
+//@ compile-flags: -Zinline-mir --crate-type=lib
+
+// EMIT_MIR inline_double_cycle.a.Inline.diff
+// EMIT_MIR inline_double_cycle.b.Inline.diff
+
+#![feature(fn_traits)]
+
+#[inline]
+pub fn a() {
+    FnOnce::call_once(a, ());
+    FnOnce::call_once(b, ());
+}
+
+#[inline]
+pub fn b() {
+    FnOnce::call_once(b, ());
+    FnOnce::call_once(a, ());
+}

